### DISCUSSION TITLE
Custom Multiselect field yielding 502 error

### DISCUSSION THREAD

**Discussion by dolphiq-allard** [craftcms/cms#12701] [Q&A]
When I try to fill my Custom Multiselectfield with options using the following code, it results in a `502 Bad Gateway` error:

```
    protected function options(): array
    {

        $userId = 116;
        $user = User::find()->id($userId)->one();

        $this->options = [];

        foreach ($user->companyTeamMembers as $teamMember) {
            $this->options[] = [
                'label' => $teamMember['memberName'],
                'value' => $teamMember['id'],
            ];
        }

        return $this->options ?? [];
    }
```

When I change from Multiselect to Dropdown the code works as expected. Checked the class reference and both fields seem to return the options in the same way. If I use a predefined array the code works as expected:

```
        $companyTeamMembers = array(
            array("id" => 1, "memberName" => "John Doe"),
            array("id" => 2, "memberName" => "Jane Smith"),
            array("id" => 3, "memberName" => "Bob Johnson"),
            array("id" => 4, "memberName" => "Alice Brown"),
        );


        $this->options = [];

        foreach ($companyTeamMembers as $teamMember) {
            $this->options[] = [
                'label' => $teamMember['memberName'],
                'value' => $teamMember['id'],
            ];
        }

        return $this->options ?? [];
    }
```

Somehow the code `$user = User::find()->id($userId)->one();` works in de Dropdown type but not in the `Multiselect` part. What am I missing here?

**Comment by brandonkelly** [comment]
Craft is probably not the thing that’s issuing that 502 error. That would generally be coming from your web server itself. You may have some sort of firewall or security extension (mod_security, Suhosin, etc.) installed that is filtering requests that look suspicious (like a SQL attack), and something about the post data is triggering it.

**Comment by dolphiq-allard** [comment]
I run the project in my local Docker environment without any hard limitations. Furthermore, if I extend the Dropdown field using the exact same code, it works like a charm so that's a bit weird.

**Reply by brandonkelly** [reply]
Try clearing out your `storage/logs/` directory and reproduce the error. Are any errors getting logged to `web*.log` or `phperrors.log`?

**Comment by dolphiq-allard** [comment]
Emptied the dir. When I use Dropdown there's just the regular logs, but as soon as I use MultiSelect I get a load of the following errors:

```
2023-02-22 11:48:50 [web.INFO] [yii\db\Connection::open] Opening DB connection: mysql:host=db;dbname=db;port=3306 {"memory":965072} 
2023-02-22 11:48:50 [web.INFO] [yii\web\Session::open] Session started {"memory":1187552} 
2023-02-22 11:48:50 [web.WARNING] [craft\elements\db\ElementQuery::prepare] Element query executed before Craft is fully initialized. {"memory":1301456} 
2023-02-22 11:48:50 [web.INFO] [delaneymethod\craftMix\CraftMix::init] Craft Mix plugin loaded {"memory":1448504} 
2023-02-22 11:48:50 [web.WARNING] [craft\elements\db\ElementQuery::prepare] Element query executed before Craft is fully initialized. {"memory":1491840} 
2023-02-22 11:48:50 [web.WARNING] [craft\elements\db\ElementQuery::prepare] Element query executed before Craft is fully initialized. {"memory":1504376} 
2023-02-22 11:48:50 [web.WARNING] [craft\elements\db\ElementQuery::prepare] Element query executed before Craft is fully initialized. {"memory":1516912} 
...
```

**Reply by brandonkelly** [reply]
That warning is saying that an element query was getting executed too early in the request, before other plugins had a chance to be initialized.

Check that none of your plugins/modules are doing anything that could trigger an element query from their `init()` methods. Instead that should wait until `Craft::$app` has triggered its `init` event.

```php
Craft::$app->on(\craft\web\Application::EVENT_INIT, function() {
    // now it's safe to execute an element query
});
```

If the plugins require Craft 4.3.5 or later (`"craftcms/cms": "^4.3.5"` in composer.json), they can use the new `onInit()` method instead:

```php
Craft::$app->onInit(function() {
    // now it's safe to execute an element query
});
```

**Comment by dolphiq-allard** [comment]
I have it like this, but it still returns that error:

```
        Craft::$app->onInit(function () {
            $this->attachEventHandlers();
            $this->_registerTwigExtensions();
            $this->_registerCustomFieldTypes();
        });
```

**Reply by brandonkelly** [reply]
There must be another plugin/module that is jumping the gun. If you have Xdebug enabled, you could set a breakpoint on that `Craft::warning()` line in `vendor/craftcms/cms/src/elements/db/ElementQuery.php` and take a look at the stack trace to see which plugin/module it’s coming from.
